File: cpp/dataStructures/basic/Clist.cpp
/** @file Clist.cpp
 *
 *  @author Jon Turner
 *  @date 2011
 *  This is open source software licensed under the Apache 2.0 license.
 *  See http://www.apache.org/licenses/LICENSE-2.0 for details.
 */
#include "Clist.h"

namespace grafalgo {

/** Constructor for Clist. 
 *  @param n1 defines the set of integers 1..n on which this object is defined
 */
Clist::Clist(int n1) : Adt(n1) { makeSpace(n()); }

/** Destructor for Clist */
Clist::~Clist() { freeSpace(); }

/** Allocate and initialize space for Clist.
 *  @param size is number of index values to provide space for
 */
void Clist::makeSpace(int size) {
	try { node = new lnode[size+1]; } catch (std::bad_alloc e) {
		stringstream ss;
		ss << "Clist::makeSpace: insufficient space for "
		   << size << "index values";
		string s = ss.str();
		throw OutOfSpaceException(s);
	}
	nn = size; clear();
}

/** Free dynamic storage used by Clist. */
void Clist::freeSpace() { delete [] node; }

/** Resize a Clist object.
 *  The old value is discarded.
 *  @param size is the size of the resized object.
 */
void Clist::resize(int size) {
	freeSpace();
	try { makeSpace(size); } catch(OutOfSpaceException e) {
		string s; s = "Clist::resize:" + e.toString(s);
		throw OutOfSpaceException(s);
	}
}

/** Expand the space available for this Clist.
 *  Rebuilds old value in new space.
 *  @param size is the size of the resized object.
 */
void Clist::expand(int size) {
	if (size <= n()) return;
	Clist old(this->n()); old.copyFrom(*this);
	resize(size); this->copyFrom(old);
}

/** Clear the data structure, moving all index values into single node lists.
 */
void Clist::clear() {
	for (index i = 0; i <= n(); i++) {
		node[i].next = node[i].prev = i;
	}
}

/** Copy into Clist from source. */
void Clist::copyFrom(const Clist& source) {
	if (&source == this) return;
	if (source.n() > n()) resize(source.n());
	else clear();
	for (index x = 0; x <= source.n(); x++) {
		node[x].next = source.suc(x);
		node[x].prev = source.pred(x);
	}
}

/** Remove an index from its list.
 *  This method turns the index into a singleton list.
 *  @param i is an index
 */
void Clist::remove(index i) {
	assert(0 <= i && i <= n());
	node[node[i].prev].next = node[i].next;
	node[node[i].next].prev = node[i].prev;
	node[i].next = node[i].prev = i;
}


/** Join two lists together.
 *  @param i is an index on some list
 *  @param j is an index on some other list
 *  Note: the method will corrupt the data structure if
 *  i and j already belong to the same list; it's the caller's
 *  responsiblity to ensure this doesn't happen
 */
void Clist::join(index i, index j) {
	assert(0 <= i && i <= n() && 0 <= j && j <= n());
	if (i == 0 || j == 0) return;
	node[node[i].next].prev = node[j].prev;
	node[node[j].prev].next = node[i].next;
	node[i].next = j; node[j].prev = i;
}

/** Produce a string representation of the object.
 *  @param s is a string in which the result will be returned
 *  @return a reference to s
 */
string& Clist::toString(string& s) const {
	index i, j; string s1;
	int *mark = new int[n()+1];
	s = "{";
	int cnt = 0;
	for (i = 1; i <= n(); i++) mark[i] = 0;
	for (i = 1; i <= n(); i++) {
		if (mark[i]) continue; 
		mark[i] = 1;
		if (node[i].next == i) continue;
		if (++cnt > 1) s += ", ";
		s += "[";
		s += Adt::item2string(i,s1);
		for (j = node[i].next; j != i; j = node[j].next) {
			mark[j] = 1;
			s += " ";
			s += Adt::item2string(j,s1);
		}
		s += "]";
	}
	s += "}";
	delete [] mark;
	return s;
}

} // ends namespace

File: java/dataStructures/basic/Clist.java
/** @file Clist.java 
 *
 * @author Jon Turner
 * @date 2011
 * This is open source software licensed under the Apache 2.0 license.
 * See http://www.apache.org/licenses/LICENSE-2.0 for details.
 */

package algoLib.dataStructures.basic;

public class Clist {
	private int N;			// list defined on ints in {1,...,N}
	private class ListNode {
	int	next;			// index of successor
	int	prev;			// index of predecessor
	};
	ListNode[] node;

	public Clist(int N) {
		this.N = N;
		node = new ListNode[N+1];
		for (int i = 0; i <= N; i++) {
			node[i] = new ListNode();
			node[i].next = node[i].prev = i;
		}
	}
	
	/** Reset the data structure, moving all items into single node lists.
	*/
	public void reset() {
		for (int i = 0; i <= N; i++) {
			node[i].next = node[i].prev = i;
		}
	}

	/** Get the successor of a list int.
	 *  @param i is a list int
	 *  @return the int that follows i in its list
	 */
	public int suc(int i) {
		assert(0 <= i && i <= N); return node[i].next;
	}
	
	/** Get the predecessor of a list int.
	 *  @param i is a list int
	 *  @return the int that precedes i in its list
	 */
	public int pred(int i) {
		assert(0 <= i && i <= N); return node[i].prev;
	}
	
	/** Remove an int from its list.
	 *  This method turns the int into a singleton list.
	 *  @param i is a list int
	 */
	public void remove(int i) {
		assert(0 <= i && i <= N);
		node[node[i].prev].next = node[i].next;
		node[node[i].next].prev = node[i].prev;
		node[i].next = node[i].prev = i;
	}
	
	/** Join two lists together.
	 *  @param i is a list int
	 *  @param j is a list int on some other list
	 *  Note: the method will corrupt the data structure if
	 *  i and j already belong to the same list; it's the caller's
	 *  responsiblity to ensure this doesn't happen
	 */
	public void join(int i, int j) {
		assert(0 <= i && i <= N && 0 <= j && j <= N);
		if (i == 0 || j == 0) return;
		node[node[i].next].prev = node[j].prev;
		node[node[j].prev].next = node[i].next;
		node[i].next = j; node[j].prev = i;
	}
	
	/** Produce a String representation of the object.
	 *  @param s is a String in which the result will be returned
	 *  @return a reference to s
	 */
	public String toString() {
		String s = "";
		int i, j;
		boolean[] mark = new boolean[N+1];
		for (i = 1; i <= N; i++) mark[i] = false;
		for (i = 1; i <= N; i++) {
			if (mark[i]) continue; 
			if (i > 1) s += ", ";
			mark[i] = true;
			s += "(";
			if (N <= 26) s += ((char) ('a'+(i-1)));
			else s += i;
			for (j = node[i].next; j != i; j = node[j].next) {
				mark[j] = true;
				s += " ";
				if (N <= 26) s = s + ((char) ('a'+(j-1)));
				else s += j;
			}
			s += ")";
		}
		return s;
	}
}
